mm2src/mm2_lib/mm2_wasm_lib.rs

```diff
@@ -0,0 +1,227 @@
+use serde_json::{self as json, Value as Json};
+use std::cell::RefCell;
+use std::sync::Mutex;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
+
+/// The errors can be thrown when using the `mm2_main` function incorrectly.
+#[wasm_bindgen]
+#[derive(Primitive)]
+pub enum Mm2MainErr {
+    AlreadyRuns = 1,
+    InvalidParams = 2,
+    NoCoinsInConf = 3,
+}
+
+impl From<Mm2MainErr> for JsValue {
+    fn from(e: Mm2MainErr) -> Self { JsValue::from(e as i32) }
+}
+
+#[derive(Deserialize)]
+struct MainParams {
+    conf: Json,
+    log_level: LogLevel,
+}
+
+impl From<MainParams> for LpMainParams {
+    fn from(orig: MainParams) -> Self { LpMainParams::with_conf(orig.conf).log_filter(orig.log_level) }
+}
+
+/// Run the MarketMaker2.
+///
+/// # Parameters
+///
+/// * `conf` is a UTF-8 string JSON.
+/// * `log_cb` is a JS function with the following signature:
+/// ```typescript
+/// function(level: number, line: string)
+/// ```
+///
+/// # Usage
+///
+/// ```javascript
+/// import init, {mm2_main, LogLevel, Mm2MainErr} from "./path/to/mm2.js";
+///
+/// const params = {
+///     conf: { "gui":"WASMTEST", mm2:1, "passphrase":"YOUR_PASSPHRASE_HERE", "rpc_password":"test123", "coins":[{"coin":"ETH","protocol":{"type":"ETH"}}] },
+///     log_level: LogLevel.Info,
+/// };
+/// let handle_log = function (_level, line) { console.log(line) };
+/// try {
+///     mm2_main(params, handle_log);
+/// } catch (e) {
+///     switch (e) {
+///         case Mm2MainErr.AlreadyRuns:
+///             alert("MarketMaker2 already runs...");
+///             break;
+///         // handle other errors...
+///         default:
+///             alert(`Unexpected error: ${e}`);
+///             break;
+///     }
+/// }
+/// ```
+#[wasm_bindgen]
+pub fn mm2_main(params: JsValue, log_cb: js_sys::Function) -> Result<(), JsValue> {
+    let params: MainParams = match params.into_serde() {
+        Ok(p) => p,
+        Err(e) => {
+            console_err!("Expected 'MainParams' as the first argument, found {:?}: {}", params, e);
+            return Err(Mm2MainErr::InvalidParams.into());
+        },
+    };
+    if params.conf["coins"].is_null() {
+        console_err!("Config must contain 'coins' field: {:?}", params.conf);
+        return Err(Mm2MainErr::NoCoinsInConf.into());
+    }
+    let params = LpMainParams::from(params);
+
+    if LP_MAIN_RUNNING.load(Ordering::Relaxed) {
+        return Err(Mm2MainErr::AlreadyRuns.into());
+    }
+    CTX.store(0, Ordering::Relaxed); // Remove the old context ID during restarts.
+
+    register_callback(WasmCallback::with_js_function(log_cb));
+    let fut = async move {
+        if LP_MAIN_RUNNING.compare_and_swap(false, true, Ordering::Relaxed) {
+            console_err!("lp_main already started!");
+            return;
+        }
+        let ctx_cb = |ctx| CTX.store(ctx, Ordering::Relaxed);
+        // use futures::FutureExt;
+        // match mm2::lp_main(params, &ctx_cb).catch_unwind().await {
```

Comment:
Do we still need this commented code?

Reply:
It's better to catch the cause of an unwinding panic if one occurs, but `FutureExt::catch_unwind` [requires](https://docs.rs/futures/0.1.31/src/futures/future/catch_unwind.rs.html#17) the future to be `safely transferrable across a catch_unwind boundary`

> `UnsafeCell<std::option::Option<std::result::Result<Vec<coins::utxo::rpc_clients::ElectrumUnspent>, JsonRpcError>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary

I think it's not necessary to catch the panic from `lp_main` because the spawn futures panic mostly.
Also, it may take some time to solve this compile error.

Reply:
Should we leave a TODO for this part then and recheck later? :slightly_smiling_face:

Reply:
Sure, I'll leave the comment